pkg/agent/controller/endpointcontroller.go

```diff
@@ -0,0 +1,445 @@
+			// TODO Change level
+			klog.Infof("ServiceImport %q updated", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			klog.Infof("ServiceImport %q deleted", key)
+			if err == nil {
+				c.serviceImportDeleted(obj, key)
+			}
+		},
+	})
+
+	go c.serviceInformer.Run(stopCh)
+	go c.runServiceImportWorker()
+
+	go c.Stop(stopCh)
+
+	return nil
+}
+
+func (c *ServiceImportController) Stop(stopCh <-chan struct{}) {
+	<-stopCh
+	c.queue.ShutDown()
+
+	klog.Infof("ServiceImport Controller stopped")
+}
+
+func (c *ServiceImportController) runServiceImportWorker() {
+	for {
+		keyObj, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for ServiceImports stopped")
+			return
+		}
+
+		key := keyObj.(string)
+
+		func() {
+			defer c.queue.Done(key)
+			obj, exists, err := c.serviceInformer.GetIndexer().GetByKey(key)
+
+			if err != nil {
+				klog.Errorf("Error retrieving the object with store is  %v from the cache: %v", c.serviceInformer.GetIndexer().ListKeys(), err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+
+				return
+			}
+
+			if exists {
+				c.serviceImportCreatedOrUpdated(obj, key)
+			} else {
+				klog.Errorf("Does not exist with key %q from the cache %v: %v", key, c.serviceInformer.GetIndexer().ListKeys(), err)
+			}
+
+			c.queue.Forget(key)
+		}()
+	}
+}
+
+func (c *ServiceImportController) serviceImportCreatedOrUpdated(obj interface{}, key string) {
+	if _, found := c.endpointController[key]; found {
+		klog.V(2).Infof("The endpointController is already running for  key %q", key)
+		return
+	}
+
+	serviceImportCreated := obj.(*lighthousev2a1.ServiceImport)
+	annotations := serviceImportCreated.ObjectMeta.Annotations
+	serviceNameSpace := annotations[OriginNamespace]
+	serviceName := annotations[OriginName]
+
+	service, err := c.kubeClientSet.CoreV1().Services(serviceNameSpace).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, err)
+		return
+	}
+
+	if service.Spec.Selector == nil {
+		klog.Errorf("The service %q in the namespace %q without selector is not supported", serviceName, serviceNameSpace)
+		return
+	}
+
+	labelSelector := labels.Set(service.Spec.Selector).AsSelector()
+	endpointInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = labelSelector.String()
+				return c.kubeClientSet.CoreV1().Endpoints(metav1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = labelSelector.String()
+				return c.kubeClientSet.CoreV1().Endpoints(metav1.NamespaceAll).Watch(options)
+			},
+		},
+		&corev1.Endpoints{},
+		0,
+		cache.Indexers{},
+	)
```

Comment:
Doesn't it share an informer since we use SharedInformer? I just wanted to listen only for the endpoints which have ServiceExport created using the label filters.